providers/fips/self_test.c

```diff
@@ -359,9 +359,13 @@ int ossl_prov_is_running(void)
                     || FIPS_state == FIPS_STATE_SELFTEST;
     static unsigned int rate_limit = 0;
 
-    if (res) {
-        rate_limit = 0;
-    } else if (FIPS_state == FIPS_STATE_ERROR) {
+    if (FIPS_state == FIPS_STATE_ERROR) {
```

Comment:
`SELF_TEST_post()` only accesses `FIPS_state` while holding the `self_test_lock` lock, even for read.  I'm wondering if the accesses here and in `ossl_set_error_state()` should also have locks.

The semantics of the rate limiting have changes but I don't think not resetting the `rate_limit` is an issue since the module can't exit the error state.

Reply:
I've complete reworked the locking in this file to address this issue. It turns out the easiest thing to do was to introduce a new lock for accessing `FIPS_state` rather than reusing `self_test_lock`. My first attempt with just `self_test_lock` was running into some deadlock issues.